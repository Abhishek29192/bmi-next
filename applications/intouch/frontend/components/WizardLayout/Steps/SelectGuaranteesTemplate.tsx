import React, { useEffect, useState } from "react";
import Select, { MenuItem } from "@bmi/select";
import { gql } from "@apollo/client";
import { useGetGuaranteeTemplatesLazyQuery } from "../../../graphql/generated/hooks";
import { GetGuaranteeTemplatesQuery } from "../../../graphql/generated/operations";
import { useWizardContext } from "../WizardContext";

const SelectGuaranteesTemplate = () => {
  const { data, setData } = useWizardContext();

  const {
    guaranteeType: { technology, coverage },
    guaranteeTemplate
  } = data;

  const [guaranteeTemplates, setGuaranteeTemplates] = useState<
    GetGuaranteeTemplatesQuery["guaranteeTemplateCollection"]["items"]
  >([]);

  const [getGuaranteeTemplates] = useGetGuaranteeTemplatesLazyQuery({
    onCompleted: ({ guaranteeTemplateCollection: { items: templates } }) => {
      if (templates.length > 0) {
        setGuaranteeTemplates(templates);
        if (!data?.guaranteeTemplate) {
          setData({
            ...data,
            guaranteeTemplate: templates[0]
          });
        }
      }
    }
  });

  const selectedTemplate = guaranteeTemplate?.sys?.id || "";

  useEffect(() => {
    getGuaranteeTemplates({
      variables: {
        technology: technology,
        coverage: coverage
      }
    });
  }, [technology, coverage]);

  const onChangeHandler = (templateId) => {
    const guaranteeTemplate = guaranteeTemplates.find(
      (item) => item.sys.id === templateId
    );
    setData({ ...data, guaranteeTemplate: guaranteeTemplate });
  };

  return (
    <div>
      {guaranteeTemplates.length > 0 && (
        <Select
          name="template"
          label="Language"
          isRequired
          style={{ margin: "10px" }}
          onChange={onChangeHandler}
          value={selectedTemplate || ""}
        >
          {guaranteeTemplates.map((template) => (
            <MenuItem value={template.sys.id} key={template.displayName}>
              {template.displayName}
            </MenuItem>
          ))}
        </Select>
      )}
    </div>
  );
};

export default SelectGuaranteesTemplate;

export const GET_GUARANTEE_TEMPLATES = gql`
  query getGuaranteeTemplates(
    $technology: String!
    $coverage: String!
    $language: String
  ) {
    guaranteeTemplateCollection(
      where: {
        coverage: $coverage
        technology: $technology
        languageCode: $language
      }
    ) {
      items {
        sys {
          id
        }
        displayName
        languageCode
        coverage
      }
    }
  }
`;
