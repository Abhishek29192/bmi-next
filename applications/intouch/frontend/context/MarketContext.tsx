import React, { useState, createContext, useEffect } from "react";
import { GetMarketsByDomainQuery } from "../graphql/generated/operations";

export type ContextProps = {
  market: GetMarketsByDomainQuery["markets"]["nodes"][0];
};

type ContextWrapperProps = ContextProps & {
  children?: React.ReactNode;
};

export const MarketContext = createContext<ContextProps>(null);
export const useMarketContext = () => React.useContext(MarketContext);

const MarketContextWrapper = ({ market, children }: ContextWrapperProps) => {
  const [currentMarket, setCurrentMarket] =
    useState<ContextProps["market"]>(market);

  useEffect(() => {
    setCurrentMarket(market);
  }, [market]);

  return (
    <MarketContext.Provider value={{ market: currentMarket }}>
      {children}
    </MarketContext.Provider>
  );
};

export default MarketContextWrapper;
