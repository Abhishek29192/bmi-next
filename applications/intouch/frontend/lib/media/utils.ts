import { GetMediaFoldersQuery } from "../../graphql/generated/operations";
import { MediaItem, MediaFolders, RootFolders, MediaTool } from "./types";

export const getRootFolders = ({
  marketContentCollection
}: GetMediaFoldersQuery): RootFolders => {
  return marketContentCollection.items[0]?.mediaLibraryRootCollection?.items;
};

export const findFolderById = (allFolders: MediaFolders, folderId: string) => {
  return allFolders.find((f) => f.sys.id === folderId);
};

export const getParentFolder = (mediaItem: MediaItem, folders: MediaFolders) =>
  folders.find((f) =>
    f.childrenCollection.items.find(
      (child) => child.sys.id === mediaItem.sys.id
    )
  );

export const getMediaItemPath = (
  mediaItem: MediaItem,
  allFolders: MediaFolders,
  rootFolders: RootFolders,
  currentPath: MediaItem[] = []
) => {
  if (!mediaItem) {
    return [];
  }
  const rootFolder = findFolderById(rootFolders, mediaItem.sys.id);

  if (rootFolder) {
    return [rootFolder, ...currentPath];
  }
  return getMediaItemPath(
    getParentFolder(mediaItem, allFolders),
    allFolders,
    rootFolders,
    [mediaItem, ...currentPath]
  );
};

const META_TYPES = {
  FOLDER: "FOLDER",
  IMAGE: "IMAGE",
  VIDEO: "VIDEO",
  EXTERNAL_LINK: "EXTERNAL_LINK",
  PDF: "PDF",
  FALLBACK: "MEDIA"
};

export const getMediaItemMeta = (mediaItem: MediaItem) => {
  if (mediaItem.__typename === "MediaFolder") {
    return META_TYPES.FOLDER;
  }
  if (mediaItem.url) {
    return mediaItem.url.includes("vimeo")
      ? META_TYPES.VIDEO
      : META_TYPES.EXTERNAL_LINK;
  }
  const contentType = mediaItem.media?.contentType;
  if (!contentType) {
    return META_TYPES.FALLBACK;
  }
  if (contentType.includes("image")) {
    return META_TYPES.IMAGE;
  }
  if (contentType.includes("pdf")) {
    return META_TYPES.PDF;
  }
  if (contentType.includes("video")) {
    return META_TYPES.VIDEO;
  }
  return META_TYPES.FALLBACK;
};

export const getMediaItemSize = (mediaItem: MediaItem) => {
  if (mediaItem.__typename !== "MediaTool" || !mediaItem.media) {
    return "";
  }
  const numBytes = formatFileSize(mediaItem.media.size);
  return numBytes;
};

// taken from StackOverflow
// https://stackoverflow.com/a/18650828/5489631
export const formatFileSize = (bytes: number) => {
  if (typeof bytes === "undefined") {
    return "";
  }
  if (bytes === 0) {
    return "0 Bytes";
  }
  const k = 1024;

  const sizes = ["Bytes", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"];

  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return [parseFloat((bytes / Math.pow(k, i)).toFixed(2)), sizes[`${i}`]].join(
    " "
  );
};

export const resizeContentfulImage = (url: string, maxWidth: number) => {
  return `${url}?w=${maxWidth}`;
};

export const getVimeoEmbedUrl = (mediaUrl: MediaTool["url"]) => {
  const videoId = mediaUrl.split("/").slice(-1)[0];
  return `https://player.vimeo.com/video/${videoId}`;
};

export const isVimeo = (mediaItem: MediaItem) => {
  return (
    mediaItem.__typename === "MediaTool" && mediaItem.url?.includes("vimeo")
  );
};

export const isExternalLink = (mediaItem: MediaItem) => {
  return (
    mediaItem.__typename === "MediaTool" &&
    mediaItem.url &&
    !mediaItem.url.includes("vimeo")
  );
};
