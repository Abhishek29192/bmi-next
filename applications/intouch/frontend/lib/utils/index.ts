import { Request } from "express";

export const formatDate = (date: string) =>
  new Intl.DateTimeFormat("en-GB", {
    dateStyle: "medium"
  } as any).format(new Date(date?.substring(0, 10)));

type MarketEnvProps = {
  currentHost: string;
  environment: string;
  market: string;
  paths: string[];
};

export const getMarketAndEnvFromReq = ({
  headers
}: Request): MarketEnvProps => {
  let market = null;
  let environment = null;

  let { host } = headers;

  const paths = host.split(".");
  const subdomain = paths[0];

  if (subdomain.indexOf("-") !== -1) {
    const parts = subdomain.split("-");
    environment = parts[0];
    market = parts[1];
  } else {
    market = subdomain;
  }

  return {
    currentHost: host,
    environment,
    market,
    paths
  };
};

export const sortArrayByField = (nodes: any[], field: string) =>
  nodes?.sort((a, b) => a?.[field]?.localeCompare(b?.[field]));
