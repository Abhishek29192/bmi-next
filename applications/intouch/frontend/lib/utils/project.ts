import { GuaranteeEventType, Project } from "@bmi/intouch-api-types";
import { GetProjectQuery } from "../../graphql/generated/operations";
import { DeepPartial } from "./types";
import {
  GuaranteeStatus,
  guaranteePrerequsitesMet,
  guaranteeSolutionGuaranteeValidate
} from "./guarantee";

export enum ProjectStatus {
  NOT_STARTED = "Not started",
  IN_PROGRESS = "In progress",
  COMPLETED = "Completed"
}

export const getProjectStatus = (startDate, endDate) => {
  const now = Date.now();
  startDate = new Date(startDate).getTime();
  endDate = new Date(endDate).getTime();

  // Start date in the future
  if (startDate > now) {
    return ProjectStatus.NOT_STARTED;
  }

  // End date in the past
  if (endDate < now) {
    return ProjectStatus.COMPLETED;
  }

  // now between start and end date
  if (startDate <= now && now <= endDate) {
    return ProjectStatus.IN_PROGRESS;
  }

  // TODO: Invalid, log it?
  return ProjectStatus.NOT_STARTED;
};

export const getProjectGuaranteeStatus = (
  project: DeepPartial<Project>
): GuaranteeStatus => {
  const guarantee = project?.guarantees?.nodes?.[0];

  // Guarantee associated with the Project - Not Applicable
  if (!guarantee) {
    return "NOT_APPLICABLE";
  }

  if (guarantee.status === "NEW") {
    return guaranteePrerequsitesMet(guarantee) ? "READY" : "STARTED";
  }

  // Other known statuses are handled in microcopy, with a fallback:
  return ["SUBMITTED", "REVIEW", "REJECTED", "APPROVED"].includes(
    guarantee.status
  )
    ? guarantee.status
    : "NOT_APPLICABLE";
};

export const getGuaranteeEventType = (
  project: DeepPartial<Project>,
  accountId: number
): GuaranteeEventType => {
  const guarantee = project?.guarantees?.nodes?.[0];

  if (!guarantee || !["SUBMITTED", "REVIEW"].includes(guarantee.status)) {
    return null;
  }

  if (guarantee.status === "SUBMITTED") return "ASSIGN_SOLUTION";
  if (guarantee.status === "REVIEW") {
    if (guarantee.reviewerAccountId !== accountId) {
      return "REASSIGN_SOLUTION";
    }
    return "UNASSIGN_SOLUTION";
  }
};
export const isProjectApprovable = (
  project: DeepPartial<Project>,
  accountId: number
): boolean => {
  const guarantee = project?.guarantees?.nodes?.[0];

  if (!guarantee) {
    return false;
  }

  return (
    guarantee.status === "REVIEW" && guarantee.reviewerAccountId === accountId
  );
};

export const checkProjectGuaranteeReview = (
  project: GetProjectQuery["project"]
): boolean => {
  const solutionGuaranteeValidationResult =
    guaranteeSolutionGuaranteeValidate(project);

  return solutionGuaranteeValidationResult.isValid;
};

export const isSolutionOrSystemGuaranteeExist = (
  project: GetProjectQuery["project"]
) => {
  return project.guarantees.nodes.some((guarantee) =>
    ["SOLUTION", "SYSTEM"].includes(guarantee.coverage)
  );
};
