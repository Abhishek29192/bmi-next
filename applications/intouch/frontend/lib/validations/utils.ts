import { get } from "lodash";
import validator from "validator";
import { ValidationResult } from "@bmi/form";

export type ValidateObjectResult = {
  result: boolean;
  missingFields: string[];
};

export const validateObject = (
  object: { [key: string]: any },
  mandatoryFields: string[]
): ValidateObjectResult => {
  const missingFields = mandatoryFields.filter(
    (mandatoryField) => !get(object, mandatoryField)
  );
  return {
    result: missingFields.length === 0,
    missingFields
  };
};

export const validateUrlInput =
  (t) =>
  (val: string): ValidationResult => {
    const isValid = !val || validator.isURL(val);
    return isValid ? null : t("common:error_messages.url_format");
  };

export const validateEmailInput =
  (t) =>
  (val: string): ValidationResult => {
    const isValid = !val || validator.isEmail(val);
    return isValid ? null : t("common:error_messages.email_format");
  };

const PHONE_NUMBER_MIN_LENGTH = 6;
const PHONE_NUMBER_MAX_LENGTH = 16;

export const validatePhoneNumberInput =
  (t) =>
  (val: string): ValidationResult => {
    const isValid =
      !val ||
      (validator.isLength(val, {
        min: PHONE_NUMBER_MIN_LENGTH,
        max: PHONE_NUMBER_MAX_LENGTH
      }) &&
        val.match(/^[\+\d]?(?:[\d-.\s()]*)$/gm)); // eslint-disable-line no-useless-escape
    return isValid ? null : t("common:error_messages.phone_number_format");
  };
