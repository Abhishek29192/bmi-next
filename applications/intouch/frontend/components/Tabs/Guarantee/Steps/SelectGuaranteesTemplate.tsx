import React from "react";
import Tooltip from "@bmi/tooltip";
import { ActionTile } from "../../../ActionTile";
import { useWizardContext } from "../../../../context/WizardContext";

export const SelectGuaranteesTemplate = () => {
  const { data } = useWizardContext();

  const templates = data.guaranteeType?.guaranteeTemplatesCollection?.items;

  return (
    <div>
      {(templates || []).map((template) => (
        <Tooltip
          title={template.displayName}
          key={template.displayName}
          style={{ padding: "5px" }}
        >
          <div>
            <ActionTile
              title={template.displayName}
              description={template.displayName}
              onClick={() => {
                // eslint-disable-next-line no-console
                console.log(template);
              }}
            />
          </div>
        </Tooltip>
      ))}
    </div>
  );
};
