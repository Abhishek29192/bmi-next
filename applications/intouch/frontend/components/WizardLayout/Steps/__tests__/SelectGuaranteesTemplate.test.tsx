import React from "react";
import { render, within } from "@testing-library/react";
import SelectGuaranteesTemplate from "../SelectGuaranteesTemplate";
import { generateProject } from "../../../../lib/tests/factories/project";
import { generateProduct } from "../../../../lib/tests/factories/product";
import { projectMembers } from "../../../../fixtures/projectMembers";
import { GetProjectQuery } from "../../../../graphql/generated/operations";
import Apollo from "../../../../lib/tests/fixtures/apollo";
import I18nProvider from "../../../../lib/tests/fixtures/i18n";
import { fireEvent, screen } from "../../../../lib/tests/utils";
import { useMarketContext } from "../../../../context/MarketContext";

const useWizardContextSpy = jest.fn();
jest.mock("../../WizardContext", () => ({
  useWizardContext: () => useWizardContextSpy()
}));
jest.mock("../../../../context/MarketContext", () => ({
  useMarketContext: jest.fn()
}));

const mockApolloData = {
  guaranteeTemplateCollection: {
    items: [
      {
        sys: {
          id: 1
        },
        languageDescriptor: "test"
      }
    ]
  }
};
jest.mock("@apollo/client", () => ({
  ...jest.requireActual("@apollo/client"),
  useLazyQuery: (_, { onCompleted }) => [
    jest.fn(() => {
      onCompleted(mockApolloData);
    })
  ]
}));

const guaranteeType = {
  evidenceCategoriesCollection: {
    items: [
      {
        name: "Waterproofing",
        minimumUploads: 1,
        description: null
      },
      {
        name: "Drainage",
        description: {
          json: {}
        }
      }
    ]
  }
};

const mockSystem: GetProjectQuery["project"]["guarantees"]["nodes"][0]["systemBySystemBmiRef"] =
  {
    id: 1,
    name: "BMI-NO-PROD-001",
    description: "",
    systemMembersBySystemBmiRef: {
      nodes: [
        {
          id: 1,
          productByProductBmiRef: generateProduct()
        }
      ]
    }
  };

const mockProject = generateProject({
  projectMembers: {
    nodes: projectMembers
  }
});

const mockData = {
  guaranteeType,
  product: generateProduct(),
  system: mockSystem,
  evidences: [
    {
      name: "test"
    }
  ]
};

describe("SelectGuaranteesTemplate", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  it("should render correct", () => {
    (useMarketContext as jest.Mock).mockImplementation(() => ({
      market: { id: 1, domain: "en" }
    }));
    useWizardContextSpy.mockReturnValue({
      project: mockProject,
      gotoNext: jest.fn(),
      gotoBack: jest.fn(),
      setData: jest.fn(),
      previousStep: 0,
      data: mockData
    });
    const { container } = render(
      <Apollo>
        <I18nProvider>
          <SelectGuaranteesTemplate />
        </I18nProvider>
      </Apollo>
    );
    expect(container).toMatchSnapshot();
    expect(useWizardContextSpy).toHaveBeenCalledTimes(2);
  });

  it("should proceed with GoBack button", () => {
    (useMarketContext as jest.Mock).mockImplementation(() => ({
      market: { id: 1, domain: "en" }
    }));
    useWizardContextSpy.mockReturnValue({
      project: mockProject,
      gotoNext: jest.fn(),
      gotoBack: jest.fn(),
      setData: jest.fn(),
      previousStep: 1,
      data: mockData
    });
    render(
      <Apollo>
        <I18nProvider>
          <SelectGuaranteesTemplate />
        </I18nProvider>
      </Apollo>
    );

    const selectCompoEl = screen.getByTestId("templates-select");
    const button = within(selectCompoEl).getByRole("button");
    fireEvent.mouseDown(button);

    const listbox = within(screen.getByRole("presentation")).getByRole(
      "listbox"
    );
    const options = within(listbox).getAllByRole("option");
    fireEvent.click(options[0]);

    expect(useWizardContextSpy).toHaveBeenCalledTimes(2);
  });
});
