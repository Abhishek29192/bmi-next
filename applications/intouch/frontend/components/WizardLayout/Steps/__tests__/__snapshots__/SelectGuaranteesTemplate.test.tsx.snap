// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectGuaranteesTemplate should render correct 1`] = `
<div>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="false"
      >
        guarantee_tab.apply_guarantee.wizard.step2.inputLabel
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Select MuiInputBase-formControl"
        data-testid="templates-select"
        style="margin: 10px;"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="mui-component-select-template"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
          id="mui-component-select-template"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          name="template"
          required=""
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-3"
          >
            <span>
              guarantee_tab.apply_guarantee.wizard.step2.inputLabel
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-required"
      />
    </div>
  </div>
</div>
`;
